internal/events/event_consumer_test.go

```diff
@@ -78,40 +69,148 @@ func Test_EventConsumer_Consume_SendDLQ(t *testing.T) {
 
 	ec.Consume(ctx)
 
 	entries := getEntries()
-	assert.Equal(t, "Waiting 2s before retrying reading new messages", entries[0].Message)
+	assert.Equal(t, "Waiting 2s before retrying handling message with key key-1", entries[0].Message)
 	assert.Equal(t, "Max backoff reached for topic test.test_topic.", entries[1].Message)
 	assert.Equal(t, "Sending message with key key-1 to DLQ for topic test.test_topic", entries[2].Message) // backoffManager.ResetBackoff() was called
 
 	consumerMock.AssertExpectations(t)
 	crashTrackerMock.AssertExpectations(t)
-	producerMock.AssertExpectations(t)
+}
+
+func Test_EventConsumer_Consume_HandleFailedOnly(t *testing.T) {
+	// setup mocks
+	consumerMock := &MockConsumer{}
+	crashTrackerMock := &crashtracker.MockCrashTrackerClient{}
+	producerMock := NewMockProducer(t)
+	failedEventHandlerMock := NewMockEventHandler(t)
+	successfulEventHandlerMock := NewMockEventHandler(t)
+
+	handlingErr := errors.New("handling message for topic test.test_topic")
+	msg := &Message{Key: "key-1", Topic: "test.test_topic"}
+
+	ec := NewEventConsumer(consumerMock, producerMock, crashTrackerMock)
+
+	// Kill the consumer after 3 seconds.
+	ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(time.Second*3))
+	defer cancel()
+
+	crashTrackerMock.
+		On("LogAndReportErrors", mock.Anything, mock.Anything, handlingErr.Error()).
+		Return()
+
+	// In the 3 seconds runtime, we're expecting call `ReadMessage` once.
+	consumerMock.
+		On("ReadMessage", ctx).Return(msg, nil).Once().
+		On("Topic").Return("test.test_topic").
+		On("Handlers").Return([]EventHandler{failedEventHandlerMock, successfulEventHandlerMock})
+
+	// In the 3 seconds runtime, we're expecting to call `Handle` twice for `FailedEventHandler`.
+	// 1 call triggered by the original message and one backoff.
+	failedEventHandlerMock.
+		On("Handle", ctx, msg).Return(handlingErr).Twice().
+		On("CanHandleMessage", ctx, msg).Return(true).
+		On("Name").Return("FailedEventHandler")
+
+	// In the 3 seconds runtime, we're expecting to call `Handle` once for `SuccessfulEventHandler`.
+	//  This is triggered by the original message. The backoff shouldn't trigger this handler.
+	successfulEventHandlerMock.
+		On("Handle", ctx, msg).Return(nil).Once().
+		On("CanHandleMessage", ctx, msg).Return(true).
+		On("Name").Return("SuccessfulEventHandler")
+
+	// We expect the message to be re-broadcasted to the same topic when consumer gets interrupted by context deadline.
+	producerMock.
+		On("WriteMessages", ctx, mock.MatchedBy(func(m []Message) bool {
+			if len(m) != 1 {
+				return false
+			}
+			message := m[0]
+			// Verify that there is only 1 successful message
+			if len(message.SuccessfulExecutions) != 1 {
+				return false
+			}
+
+			// Verify that one error is recorded correctly. in 3 seconds there should be one backoff.
+			if len(message.Errors) != 1 && message.Errors[0].ErrorMessage != handlingErr.Error() {
+				return false
+			}
```

Comment:
It's a bit hard to figure out what failed when you rely on booleans... 

You can return `assert.Equal(...) here, since it returns a boolean.

Reply:
This one is a matter of preference, so no need to do any changes on your code: I actually prefer to use `.Run` rather than `MatchedBy` because matchedBy relies on reflection, which is a bit buggy for complex objects:

https://github.com/stellar/stellar-disbursement-platform-backend/blob/cbee8598935546282671b0b60a88b2c1913fa342/internal/transactionsubmission/transaction_worker_test.go#L1889-L1908

Reply:
That's a great suggestion. You're right that booleans are not practical for asserting. 
Addressed in db76b256de5ccb95eb9ec5ecfb37bbb0675fd044